### PR TITLE
Report: Add zoom in feature for ramp charts

### PR DESCRIPTION
**Pull request by ongspxm**
- [x] highlighting of ramp charts #544 
- [x] refactoring ramp charts #572 
- [x] refactoring tab display (from master) #566 
- [x] add mechanism to load diff tabs #574 
- [x] the actual zoomin tab #575

```
The commit granularity setting for the ramp chart slice produces a ramp 
chart that may be too dense to reveal any substantial information. 

Let's add a side panel that provides a zoomed-in view of the ramp 
chart, selected based on a specified time frame.
```

### CODE REVIEW

**Comment by myteo** [comment]
Oops sorry. Accidentally clicked the checkbox😅

**Comment by emer7** [comment]
Eh how to activate the zoom in ah?

**Comment by ongspxm** [comment]
> Eh how to activate the zoom in ah?

shift left click on the ramp chart

File: frontend/src/index.pug [hunk 1]
```diff
@@ -39,14 +41,19 @@ html
...
         #tabs-wrapper(v-if="isTabActive")
           #tab-resize(onmousedown="registerMouseMove()")
-          .tab-close(v-on:click="isTabActive=false")
+          .tab-close(v-on:click="isTabActive=false;")
             i.fas.fa-caret-right
           .tab-content
             .tab-panes
               v_authorship#tab-authorship.tab-pane(
-                v-if="tabType === 'authorship'"
+                v-if="tabType=='authorship'"
```
Comment (chelseyong) [inline frontend/src/index.pug]: Hmm why change?

**Comment by emer7** [comment]
When I tried to zoom in at particularly crowded area, I will instead click on the ramp + open at new window 😕

**Comment by ongspxm** [comment]
test

**Comment by emer7** [comment]
image
this quite hard 😢

**Comment by ongspxm** [comment]
> image
> this quite hard cry

issued a fix, hahah try again

**Comment by ongspxm** [comment]
all ready for review

**Comment by ongspxm** [comment]
image

cos you have a lot of small commits on that day, so they lump into one bunch, make one big green triangle

**Comment by eugenepeh** [comment]
>cos you have a lot of small commits on that day, so they lump into one bunch, make one big green triangle
>
>if you change into commit view can see the difefrence

I see, the zooming for now is just always commit view?

**Comment by ongspxm** [comment]
> > cos you have a lot of small commits on that day, so they lump into one bunch, make one big green triangle
> > if you change into commit view can see the difefrence
> 
> I see, the zooming for now is just always commit view?

yep, that was the orginal intent, no? zoom into the hairy commit view

**Comment by ongspxm** [comment]
@yamidark added in the user guide

**Comment by ongspxm** [comment]
image

oh noes, java8 cannot install properly

**Comment by yamidark** [comment]
>oh noes, java8 cannot install properly

Resolved.

**Comment by ongspxm** [comment]
@yamidark nice :+1:

**Comment by fzdy1914** [comment]
image
Some texts are selected wrongly.

**Comment by ongspxm** [comment]
@fzdy1914 how do you replicate the error? Just shift click shift click?

**Comment by fzdy1914** [comment]
a

**Comment by ongspxm** [comment]
the `reset date range` is for when it zoomed in, then it will zoom out everything.

what browser you using for the gif one? is it you keep holding shift?

**Comment by fzdy1914** [comment]
I am using Chrome. And I am not holding shift.

**Comment by ongspxm** [comment]
oh yeah, this does seems to be a problem on chrome

**Comment by ongspxm** [comment]
i tried and it seems like ctrl key works, im gonna changed to ctrl instead

**Comment by ongspxm** [comment]
ready to merge? @yamidark @eugenepeh

File: frontend/src/static/css/style.scss [hunk 2]
```diff
@@ -233,27 +238,21 @@ header{
...
+            height:100%;
             position:absolute;
-            width:0;
-            height:0;
+            top:0;
 
-            border-left-color: rgba(0,0,0,0);
-            border-left-style: solid;
+            &.show{
+                background-color:#fffa;
```
Comment (eugenepeh) [inline frontend/src/static/css/style.scss]: how come not using `mui-color` for this?

File: frontend/src/static/css/style.scss [hunk 3]
```diff
@@ -276,6 +275,58 @@ header{
...
+        &--color0{ border-bottom:$height rgba(mui-color("orange"), 0.5) solid;  }
+        &--color1{ border-bottom:$height rgba(mui-color("light-blue"), 0.5) solid;  }
+        &--color2{ border-bottom:$height rgba(mui-color("green"), 0.5) solid;  }
+        &--color3{ border-bottom:$height rgba(mui-color("indigo"), 0.5) solid;  }
+        &--color4{ border-bottom:$height rgba(mui-color("pink"), 0.5) solid;  }
+    }
+}
+
+/* Ramp */
+.ramp{
```
Comment (eugenepeh) [inline frontend/src/static/css/style.scss]: why is there duplicated code?

File: frontend/src/static/js/main.js [hunk 4]
```diff
@@ -160,21 +160,33 @@ window.app = new window.Vue({
...
+      this.isCollapsed = false;
+      this.tabType = tabName;
     },
 
     updateTabAuthorship(obj) {
-      this.deactivateTab();
       this.tabInfo.tabAuthorship = Object.assign({}, obj);
+      this.activateTab('authorship');
+    },
+    updateTabZoomin(obj) {
```
Comment (eugenepeh) [inline frontend/src/static/js/main.js]: zoom in? zooming?

File: frontend/src/static/js/v_authorship.js [hunk 5]
```diff
@@ -38,10 +38,12 @@ window.vAuthorship = {
       filesBlankLinesObj: {},
       totalLineCount: '',
       totalBlankLineCount: '',
+
       filesSortType: 'lineOfCode',
-      toReverseSortFiles: false,
       activeFilesCount: 0,
       filterSearch: '*',
+
+      toReverseSortFiles: false,
```
Comment (eugenepeh) [inline frontend/src/static/js/v_authorship.js]: changes on this file seems irrelevant

File: frontend/src/zoomin.pug [hunk 6]
```diff
@@ -0,0 +1,19 @@
+#zoomin
+  h1 {{ info.user.displayName }} ({{ info.user.name }})
```
Comment (eugenepeh) [inline frontend/src/zoomin.pug]: for pug files that belongs to a sub component, e.g. side window, can we place them in a subfolder?

File: frontend/src/static/js/main.js (hunk 4 above)
Comment (ongspxm) [reply frontend/src/static/js/main.js]: haha this is right, tab "zoom in"

File: frontend/src/zoomin.pug (hunk 6 above)
Comment (ongspxm) [reply frontend/src/zoomin.pug]: i wanna move summary and authorship out as well, i think i will do those outside, imma gonna do for this first

**Comment by ongspxm** [comment]
ready for a review

File: frontend/src/static/js/main.js (hunk 4 above)
Comment (eugenepeh) [reply frontend/src/static/js/main.js]: If it is suppose to be "zoom in" then capitalize the `I`.

File: frontend/src/index.pug [hunk 7]
```diff
@@ -41,14 +43,19 @@ html
 
         #tabs-wrapper(v-if="isTabActive")
           #tab-resize(onmousedown="registerMouseMove()")
-          .tab-close(v-on:click="isTabActive=false")
+          .tab-close(v-on:click="isTabActive=false;")
```
Comment (eugenepeh) [inline frontend/src/index.pug]: doesnt seems like a necessary change for this PR; other lines are not practicing this.

File: frontend/src/index.pug [hunk 8]
```diff
@@ -27,8 +27,10 @@ html
         #summary-wrapper
           .summary-status(v-if="loadedRepo<repoLength") Loading repos ... {{ loadedRepo }} / {{ repoLength }}
           v_summary(
+            ref="summary",
             v-bind:repos="getUsers()",
-            v-on:view-authorship="updateTabAuthorship",
+            v-on:view-zoomin="updateTabZoomin",
+            v-on:view-authorship="updateTabAuthorship"
```
Comment (eugenepeh) [inline frontend/src/index.pug]: is the comma, meant to be missing?

File: frontend/src/ramp.pug [hunk 9]
```diff
@@ -0,0 +1,34 @@
+.ramp
+  template(v-if="tframe === 'commit'")
+      template(v-for="(slice, j) in user.commits")
```
Comment (eugenepeh) [inline frontend/src/ramp.pug]: why is this double indented?

File: frontend/src/static/css/style.scss [hunk 10]
```diff
@@ -9,6 +9,10 @@ vuetemplate{
     display:none;
 }
 
+a{
+  cursor:pointer;
```
Comment (eugenepeh) [inline frontend/src/static/css/style.scss]: 4 space

File: frontend/src/static/js/v_summary.js [hunk 11]
```diff
@@ -9,9 +9,71 @@ window.comparator = (fn) => function compare(a, b) {
...
+}
+
+window.viewClick = function viewClick(evt) {
+  if (drags.length === 2) {
+    drags = [];
+  }
+
+  if (evt.ctrlKey) {
+    return drags.length === 0
+      ? dragViewDown(evt)
```
Comment (eugenepeh) [inline frontend/src/static/js/v_summary.js]: double indent

File: frontend/src/tabs/zoomin.pug [hunk 12]
```diff
@@ -0,0 +1,19 @@
+#zoomin
+  h1 {{ info.user.displayName }} ({{ info.user.name }})
+  a(v-on:click="openSummary") Show ramp chart for this period
+
+  h2 {{ info.user.repoName }}
+  v_ramp(
+    v-bind:user="info.user",
```
Comment (eugenepeh) [inline frontend/src/tabs/zoomin.pug]: double indent

**Comment by eugenepeh** [comment]
@ongspxm in case you missed out this comment
https://github.com/reposense/RepoSense/pull/573#pullrequestreview-232563759
```
I think call it tabZoom is intuitive enough, since we're not going to have a ZoomOut, no need to emphasize it by adding an In.
```

File: frontend/src/static/js/v_summary.js (hunk 11 above)
Comment (eugenepeh) [reply frontend/src/static/js/v_summary.js]: doesnt seems to be resolved.

File: frontend/src/static/js/v_summary.js [hunk 13]
```diff
@@ -9,9 +9,71 @@ window.comparator = (fn) => function compare(a, b) {
   return 1;
 };
 
+// ui funcs, only allow one ramp to be highlighted //
+let drags = [];
+
+function getBaseTarget(target) {
+  return (target.className === 'summary-chart__ramp')
+    ? target : getBaseTarget(target.parentElement);
```
Comment (eugenepeh) [inline frontend/src/static/js/v_summary.js]: same for here, needs another 2 space of indentation

Comment (ongspxm) [reply frontend/src/static/js/v_summary.js]: oh yah, these are the one that was complained about in the eslint, i changed them once, then i saw the eslint got error.

Comment (ongspxm) [reply frontend/src/static/js/v_summary.js]: https://travis-ci.org/reposense/RepoSense/jobs/527111802

Comment (eugenepeh) [reply frontend/src/static/js/v_summary.js]: seems like eslint needs to be updated with this rule then 😛

Comment (ongspxm) [reply frontend/src/static/js/v_summary.js]: i do that in this pr?

Comment (eugenepeh) [reply frontend/src/static/js/v_summary.js]: can do it towards the master, then fix it by rebase or when merging

Comment (ongspxm) [reply frontend/src/static/js/v_summary.js]: eh... i checked the memberexpression already is 2, i think it doesn't really recognize the `?` `:` if else clause as a parameter.

the other choice is to make everything one line, but its kinda not very nice to read, i keep the current style? cos it doesn't really make sense to change it to the full if else statement either.

Comment (eugenepeh) [reply frontend/src/static/js/v_summary.js]: that is probably a different rule

Comment (eugenepeh) [reply frontend/src/static/js/v_summary.js]: should be this one https://eslint.org/docs/rules/indent#flatternaryexpressions

Comment (ongspxm) [reply frontend/src/static/js/v_summary.js]: doesn't do double indent one worh, is it okay if i on this, put it to a new line, but only one indent?

Comment (eugenepeh) [reply frontend/src/static/js/v_summary.js]: can you try `"ConditionalExpression": 2`?

Comment (ongspxm) [reply frontend/src/static/js/v_summary.js]: image

this was the only reference to conditionalexpression leh

Comment (eugenepeh) [reply frontend/src/static/js/v_summary.js]: yes, just think that we should test whether if it is an undocumented feature. It doesn't quite make sense to add this to ignore nodes but not to the indentation.

If it doesn't work for indentation, we can add the `ignoreNodes` rule anyway so that it can adhere to our standard.

**Comment by ongspxm** [comment]
as in, i tried, didn't work, gave some error, so going to use ignoreNode
instead.

**Comment by ongspxm** [comment]
all done :+1: ready @eugenepeh

**Comment by eugenepeh** [comment]
@ongspxm can you resolve the conflict?

**Comment by ongspxm** [comment]
@yong24s merge pls? hahah

**Comment by yong24s** [comment]
@ongspxm 

Fix the conflict then we'll do one final review 👍
